pkg/skaffold/build/cluster/kaniko.go

```diff
@@ -110,3 +123,89 @@ func (b *Builder) copyKanikoBuildContext(ctx context.Context, workspace string,
 
 	return nil
 }
+
+func evaluateEnv(env []v1.EnvVar, additional ...v1.EnvVar) ([]v1.EnvVar, error) {
+	// Prepare additional envs
+	addEnv := make(map[string]string)
+	for _, addEnvVar := range additional {
+		addEnv[addEnvVar.Name] = addEnvVar.Value
+	}
+
+	// Evaluate provided env variables
+	var evaluated []v1.EnvVar
+	for _, envVar := range env {
+		val, err := util.ExpandEnvTemplate(envVar.Value, nil)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get value for env variable %q: %w", envVar.Name, err)
+		}
+
+		evaluated = append(evaluated, v1.EnvVar{Name: envVar.Name, Value: val})
+
+		// Provided env variables have higher priority than additional (generated) ones
+		delete(addEnv, envVar.Name)
+	}
+
+	// Append additional (generated) env variables
+	for name, value := range addEnv {
+		if value != "" {
+			evaluated = append(evaluated, v1.EnvVar{Name: name, Value: value})
+		}
+	}
+
+	return evaluated, nil
+}
+
+func envMapFromVars(env []v1.EnvVar) map[string]string {
+	envMap := make(map[string]string)
+	for _, envVar := range env {
+		envMap[envVar.Name] = envVar.Value
+	}
+	return envMap
+}
+
+func generateEnvFromImage(imageStr string) ([]v1.EnvVar, error) {
+	repoStr, nameStr, tagStr, err := parseImageParts(imageStr)
+	if err != nil {
+		return nil, err
+	}
+	var generatedEnvs []v1.EnvVar
+	generatedEnvs = append(generatedEnvs, v1.EnvVar{Name: "IMAGE_REPO", Value: repoStr})
+	generatedEnvs = append(generatedEnvs, v1.EnvVar{Name: "IMAGE_NAME", Value: nameStr})
+	generatedEnvs = append(generatedEnvs, v1.EnvVar{Name: "IMAGE_TAG", Value: tagStr})
+	return generatedEnvs, nil
+}
+
+func parseImageParts(imageStr string) (string, string, string, error) {
```

Comment:
I think you should be able to use [`docker.ParseReference`](https://godoc.org/github.com/containers/image/docker#ParseReference) here instead. we do this a few other places in the code

Reply:
That's how I started out until I realized that `docker.ParseReference ` can't split out the image name from the repo path, which we need for our CI setup. 

I could replace part of the function code with a call to `docker.ParseReference` and only leave the code to split IMAGE_NAME from IMAGE_REPO if you prefer.

What do you think @nkubala ?

Reply:
sorry I totally missed this one. I was actually wrong on my original comment, we're using docker's (`reference.Parse`)[https://godoc.org/github.com/docker/distribution/reference#Parse] in skaffold. we finesse that into a `ImageReference` struct which should be able to split image name and repo path (we call it `Domain`) right?

if it isn't doing everything you need, I think that code should live somewhere near here so we can embed those pieces into the `ImageReference` struct.

Reply:
@nkubala no worries. I'm not sure that it would make sense to refactor the ImageReference to change the way it splits image FQDN's since I assume there was a reason for this specific split.

However, I refactored my code to use docker.ParseReference (which returns the ImageReference you pointed out) instead of implementing it from scratch. Was that what you had in mind?

Reply:
hey @DanielSel, I dug in a little more into `ParseReference` and how it creates the `ImageReference`. take the image `gcr.io/k8s-skaffold/skaffold:foo` as an example:

BaseName -> `gcr.io/k8s-skaffold/skaffold`
Domain -> `gcr.io`
Path -> `k8s-skaffold/skaffold`

the part you're interested in here is the "Repo", which would be `gcr.io/k8s-skaffold`. I think part of the confusion is that our terminology is not quite right here: `BaseName` is really "Unqualified Image Name" (to me at least), which is also "Repo" + "Image Name" (both of which are missing from this struct).

I propose that you take your logic here to split the image parts (which seems mostly fine to me as is) and move it into `docker.ParseReference`, and then embed "Repo" and "Image Name" into the `ImageReference` struct we already have. this won't change the existing usage of the struct around the codebase, but could be used later on if we wanted to, and is easier to find if we ever want to make changes to it. this isn't kaniko-specific logic, so it doesn't really belong in the kaniko package. WDYT?

Reply:
(I think) I got you now :)
Code updated.